File: xxf_view/src/main/java/com/xxf/view/utils/RAUtils.java
package com.xxf.view.utils;

import android.view.MotionEvent;
import android.view.View;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * @author xuanyouwu
 * @email xuanyouwu@163.com
 * @time 2016-11-10 17:50
 * 防爆点击
 */

public class RAUtils {
    /**
     * 记录一个路由上次跳转的时间
     */
    private static final Map<String, Long> routerJumpRecordLastTimes = new ConcurrentHashMap<>();

    private RAUtils() {
    }

    public static final long DURATION_DEFAULT = 500;

    public static boolean isLegalDefault() {
        return RAUtils.isLegal(RAUtils.DURATION_DEFAULT);
    }

    /**
     * 防爆 阻力 false 表示暴力点击
     *
     * @param duration 点击间隔
     * @return
     */
    public static boolean isLegal(long duration) {
        return isLegal(RAUtils.class.getName(), duration);
    }

    public static boolean isLegal(String path, long duration) {
        Long lastRecord = routerJumpRecordLastTimes.get(path);
        long current = System.currentTimeMillis();
        routerJumpRecordLastTimes.put(path, current);
        return lastRecord == null || System.currentTimeMillis() - lastRecord >= duration;
    }


    public static boolean inRangeOfView(View view, MotionEvent ev) {
        int[] location = new int[2];
        view.getLocationOnScreen(location);
        int x = location[0];
        int y = location[1];
        if (ev.getX() < x || ev.getX() > (x + view.getWidth()) || ev.getY() < y || ev.getY() > (y + view.getHeight())) {
            return false;
        }
        return true;
    }
}

File: lib_utils/src/main/java/com/xxf/utils/RAUtils.kt
package com.xxf.utils

import android.os.SystemClock
import java.util.WeakHashMap

/**
 * @author xuanyouwu
 * @email xuanyouwu@163.com
 * @time 2016-11-10 17:50
 * 防爆点击，截流工具
 */
object RAUtils {
    /**
     * 记录上次生效的时间
     */
    private val effectedMap: WeakHashMap<String, Long> = WeakHashMap()
    const val DURATION_DEFAULT: Long = 500

    /**
     * 阀门是否有效
     * @return 是否有效;true表示应该是合法的,可以通过
     */
    val isLegalDefault: Boolean
        get() = isLegal(DURATION_DEFAULT)

    /**
     * 阀门是否有效
     * @param duration 阀门区间 单位 毫秒
     * @return 是否有效;true表示应该是合法的,可以通过
     */
    fun isLegal(duration: Long): Boolean {
        return isLegal(RAUtils::class.java.name, duration)
    }

    /**
     * 阀门是否有效
     * @param key 唯一标识阀门钥匙
     * @param duration 阀门区间 单位 毫秒
     * @return 是否有效;true表示应该是合法的,可以通过
     */
    fun isLegal(key: String, duration: Long): Boolean {
        val lastEffected = effectedMap.getOrPut(key) { 0L }
        val current = SystemClock.elapsedRealtime()
        if (current - lastEffected >= duration) {
            effectedMap[key] = current
            return true
        }
        return false
    }
}